flight_planning/v1/user_notification.yaml

```diff
@@ -0,0 +1,20 @@
+info:
+  title: Definitions used in flight planning interface for user notification.
+components:
+  schemas:
+    UserNotification:
+      type: object
+      description: Notification sent to user.
+      required:
+        - notification_sent_at
+      properties:
+        notification_sent_at:
```

Comment:
We were very careful with these requirements to have them be when the service SENDS the notification... not when anything receives them or "observes" them because that is now a requirement on another system or process, not the USS-under-test or the service-under-test.  Is "observation time" somewhere in the requirements? Agree with Punam that `notification_sent_at` helps provide a measure that is appropriate to the requirement.

Reply:
> We are getting this info from a USS endpoint about the user notifications.

Not necessarily; that's what I'm saying in my explanation above -- this is an endpoint that provides information about user notifications.  There are lots of potential ways to implement that kind of endpoint, and actually there are multiple entities who could potentially provide it.  Let's think back to manual tests again.  In UFT, how did the test director determine if something showed up in a USS's UI?  The actual configuration that was used was that a MAAP employee operating an instance of the USS's software reported to the test director what they saw.  So, the test director did not interact directly with any USS to get this information.

Getting user notifications falls into [the first category of interactions in this API](https://github.com/interuss/automated_testing_interfaces/tree/main/flight_planning#overview): _"The test director is instructing a USS representative to use the USS system normally. In this case, it could be that the USS representative is not from the USS company at all, but merely a third party contracted to assist with the test by using the USS's system as they were trained."_  The [design doc](https://github.com/interuss/tsc/blob/main/investigations/user_notification/F3548-21_SCD009x.pdf) specifies that we're talking about _"the automated equivalent to the orange exchange between the test director and and green user above"_ (with a similar diagram as mentioned here).

An ideal way to implement this API would be to have a physical robot that operated a mouse on a laptop (or tapped a screen on a phone) in response to requests to plan flights, and used computer vision to look at the laptop screen or phone to determine the results of those attempts and what user notifications were observed.  The USS would not need to be the one who implements this robot.

The test director is not necessarily talking directly to the USS when it accesses the flight_planning API.  This is critically important and I've tried to explain it many different ways.  Please let me know if we need to discuss further in real time or what is unclear about the dotted green scope box in the diagram above.

> We were very careful with these requirements to have them be when the service SENDS the notification... not when anything receives them or "observes" them because that is now a requirement on another system or process, not the USS-under-test or the service-under-test.

Yes, that is correct -- the F3548-21 requirement that we would like to test by evaluating the information collected by this tool/sensor pertains (intentionally) only to a USS's dispatch of the notification, so therefore observation involves additional systems (notification transit, reception, processing, and display).  However, this PR already takes care of that reasonably well when it says, _"The user notifications should be available to be queried within 5 seconds of notifying the user."_  Yes, this is an additional requirement, but it is [reasonably necessary to conduct tests](https://github.com/interuss/monitoring/blob/main/monitoring/uss_qualifier/automated_testing.md#necessity) in the same way as implementing the flight_planning API in general is an additional requirement but reasonably necessary to conduct tests.

Again, consider the manual analog: UFT is conducting tests and a conflict is supposed to pop up on USS X's screen, but doesn't.  The test director wants to mark USS X as failing the requirement to inform the operator of the conflict, but USS X objects that they're compliant with the notification dispatch requirement, it's just that the notification didn't get to the display so the test director can't mark them as failing.  I think that argument should work about as well here as it would there.  "Your system must produce observable notifications" seems like an extremely reasonable requirement to impose on a USS for the purpose of enabling practical testing which needs to observe those notifications.

When we're creating these tests, [we want extremely high specificity](https://github.com/interuss/monitoring/blob/main/monitoring/uss_qualifier/automated_testing.md#failure) -- if our tests incorrectly indicate that a USS has failed to meet requirements, then they can't be used for the intended purpose of substituting for a manual checkout since "fail" doesn't really mean "fail".  We want our sensitivity to be as high as practical, but it certainly does not need to be perfect.  In manual tests, USSs that actually fail to meet requirements are sometimes not detected as failing requirements -- this is because it is impossible to conduct all possible tests that would detect all possible non-compliance.  Similarly, that is ok here -- we try to detect all non-compliance that it is practical to detect, but there is no expectation of perfect sensitivity.  And, we do not want to sacrifice the usability of the tests to gain a non-critical amount of sensitivity -- when a verbal response produces 80% sensitivity, we do not want to require drilling holes into the patient's skull to insert neural probes that will produce 90% sensitivity.  So, when we wait 10 seconds for a missing user notification, it's true that we will likely fail to declare a USS which takes 6 seconds to dispatch a notification (and therefore does not comply with the requirement to dispatch within 5 seconds) as non-compliant.  I believe that is fine for today.  If it becomes important to differentiate between USSs that take 4 seconds and USSs that take 6 seconds, then we can investigate ways to improve the tests at that time, but for now I think we're more focused on getting to full basic coverage.  My understanding of the intent of the 5 second deadline was more than we need a deadline at which we're allowed to declare failure rather than we actually expect USSs to take that long -- notifications are expected promptly and 5 seconds is the standard bar for "not prompt".  So, "exactly 5 seconds" is not the important validation; rather "sent the notification" is the important validation.

> Agree with Punam that notification_sent_at helps provide a measure that is appropriate to the requirement.

The problem is that the person in the green shirt generally does not have this information.  The information they do have is when they observed the notification.  That is plenty sufficient to conduct effective manual checkouts so it should be sufficient here.  Respondents [should not be expected to provide information that they do not have](https://github.com/interuss/monitoring/blob/main/monitoring/uss_qualifier/automated_testing.md#automated-testing-interfaces) and someone responding a flight_planning query [does not necessarily have access to the USS's internal systems](https://github.com/interuss/automated_testing_interfaces/tree/main/flight_planning#overview).